docs/architecture/adr-031-msg-service.md

```diff
@@ -0,0 +1,222 @@
+(`MsgSubmitProposal` in this case). 
+
+### Decoding
+
+When decoding, `TxBody.UnpackInterfaces` will need a special case
+to detect if `Any` type URLs match the service method format (ex. `/cosmos.gov.Msg/SubmitProposal`)
+by checking for two `/` characters. Messages that are method names plus request parameters
+instead of a normal `Any` messages will get unpacked into the `ServiceMsg` struct:
+ 
+```go
+type ServiceMsg struct {
+  // MethodName is the fully-qualified service name
+  MethodName string
+  // Request is the request payload
+  Request MsgRequest
+}
+```
+
+### Routing
+
+In the future, `service` definitions may become the primary method for defining
+`Msg`s. As a starting point, we need to integrate with the SDK's existing routing
+and `Msg` interface. 
+
+To do this, `ServiceMsg` implements the `sdk.Msg` interface and its handler does the
+actual method routing, allowing this feature to be added incrementally on top of
+existing functionality.
+
+### `MsgRequest` interface
+
+All request messages will need to implement the `MsgRequest` interface which is a
+simplified version of `Msg`, without `Route()`, `Type()` and `GetSignBytes()` which
+are no longer needed:
+
+```go
+type MsgRequest interface {
+  proto.Message
+  ValidateBasic() error
+  GetSigners() []AccAddress
+}
+```
+
+`ServiceMsg` will forward its `ValidateBasic` and `GetSigners` methods to the `MsgRequest`
+methods.
+
+### Module Configuration
+
+In [ADR 021](./adr-021-protobuf-query-encoding.md), we introduced a method `RegisterQueryService`
+to `AppModule` which allows for modules to register gRPC queriers.
+
+To register `Msg` services, we attempt a more extensible approach by converting `RegisterQueryService`
+to a more generic `RegisterServices` method:
+
+```go
+type AppModule interface {
+  RegisterServices(Configurator)
+  ...
+}
+
+type Configurator interface {
+  QueryServer() grpc.Server
+  MsgServer() grpc.Server
+}
+
+// example module:
+func (am AppModule) RegisterServices(cfg Configurator) {
+	types.RegisterQueryServer(cfg.QueryServer(), keeper)
+	types.RegisterMsgServer(cfg.MsgServer(), keeper)
+}
+```
+
+The `RegisterServices` method and the `Configurator` interface are intended to
+evolve to satisfy the use cases discussed in [\#7093](https://github.com/cosmos/cosmos-sdk/issues/7093)
+and [\#7122](https://github.com/cosmos/cosmos-sdk/issues/7421).
+
+When `Msg` services are registered, the framework _should_ verify that all `Msg...Request` types
+implement the `MsgRequest` interface described above and throw an error during initialization rather
+than later when transactions are processed.
+
+### `Msg` Service Implementation
+
+Just like query services, `Msg` service methods can retrieve the `sdk.Context`
+from the `context.Context` parameter method using the `sdk.UnwrapSDKContext`
+method:
+
+```go
+package gov
+
+func (k Keeper) SubmitProposal(goCtx context.Context, params *types.MsgSubmitProposal) (*MsgSubmitProposalResponse, error) {
```

Comment:
So basically, we define the service file and protobuf messages, and add a few lines of boilerplate (to register the services), and we can just then focus on writing the handler/querier methods?

Sounds like this removes a lot of the tedious work from module creation.

Reply:
That's the idea 👍